tests/test.html.js

```diff
@@ -0,0 +1,128 @@
+    var contents = validHTML();
+    var htmlScanner = new HTMLScanner(contents, 'index.html');
+
+    return htmlScanner.scan()
+      .then((validatorMessages) => {
+        assert.equal(validatorMessages.length, 0);
+      });
+  });
+
+  it('should handle unicode characters', () => {
+    var contents = validHTML('<strong>🎉</strong>');
+    var htmlScanner = new HTMLScanner(contents, 'index.html');
+
+    return htmlScanner.scan()
+      .then((validatorMessages) => {
+        assert.equal(validatorMessages.length, 0);
+      });
+  });
+
+  it('should require <prefwindow> tag to have an id attribute', () => {
+    var badHTML = validHTML(singleLineString`<prefwindow
+      xmlns="http://www.mozilla.org/keymaster/gatekeeper/there.is.only.xul"
+      title="My application: configuration"
+      onunload="onUnload(event.target)">
+    </prefwindow>`);
+    var goodHTML = validHTML(singleLineString`<prefwindow
+      xmlns="http://www.mozilla.org/keymaster/gatekeeper/there.is.only.xul"
+      id="my-config-dialog"
+      title="My application: configuration"
+      onunload="onUnload(event.target)">
+    </prefwindow>`);
+    var htmlScanner = new HTMLScanner(badHTML, 'index.html');
+
+    return htmlScanner.getHTMLDoc()
+      .then(($) => {
+        return rules.ensureRequiredAttributes($, htmlScanner.filename);
+      })
+      .then((validatorMessages) => {
+        assert.equal(validatorMessages.length, 1);
+        assert.equal(validatorMessages[0].code,
+                     messages.PREFWINDOW_REQUIRES_ID.code);
+        assert.equal(validatorMessages[0].type, VALIDATION_ERROR);
+
+        // Make sure there are no errors when an ID is provided.
+        htmlScanner = new HTMLScanner(goodHTML, 'index.html');
+        return htmlScanner.getHTMLDoc();
+      })
+      .then(($) => {
+        return rules.ensureRequiredAttributes($, htmlScanner.filename);
+      })
+      .then((validatorMessages) => {
+        assert.equal(validatorMessages.length, 0);
+      });
+  });
+
+  it('should not blow up when handed malformed HTML', () => {
+    var html = validHTML('<div>Howdy <!-- >');
+    var htmlScanner = new HTMLScanner(html, 'index.html');
+
+    return htmlScanner.scan()
+      .catch(() => {
+        assert.fail(null, null, 'Unexpected error');
+      });
+  });
+
+  it('should return an already-parsed htmlDoc if exists', () => {
+    var contents = validHTML();
+    var htmlScanner = new HTMLScanner(contents, 'index.html');
+
+    sinon.spy(cheerio, 'load');
+
+    return htmlScanner.getHTMLDoc()
+      .then(() => {
+        return htmlScanner.getHTMLDoc();
+      })
+      .then(() => {
+        assert.ok(cheerio.load.calledOnce);
+      });
+  });
+
+  it('should run all rules in rules/html', () => {
+    var contents = validHTML();
+    var htmlScanner = new HTMLScanner(contents, 'index.html');
+    var fakeRule = () => {
+      Promise.resolve([]);
+    };
+    var fakeRules = {
+      iAmAFakeRule: fakeRule,
```

Comment:
Yeah the interface for stubs is pretty much the same as spys on the wholle. Spys are generally more useful for wrapping a real func so you can know it was called. If you just want a stand-in then stubs are the better way to go.